syntax = "proto3";

package chat;

service Chat {
	rpc ChatStream(stream StreamRequest) returns  (stream StreamResponse) {}
}

message StreamRequest {
    oneof RequestEvent {
        LoginRequest login_request = 1;
		Message message_request  = 2;
        ListAccountsRequest listAccounts_request = 3;
        DeleteAccountRequest deleteAccount_request = 4;
        LogoutRequest logout_request = 5;
    }
}

message StreamResponse {
     oneof ResponseEvent {
        LoginResponse login_response = 1;
        Message message_response = 2;
		ListAccountsResponse listAccounts_response = 3;
        DeleteAccountResponse deleteAccount_response  = 4;
        LogoutResponse logout_response = 5;
    }
}

message Message {
    string username = 1;
    string message = 3;
}

message ListAccountsRequest {
    string username = 1;
    string wildcard = 2;
}

message DeleteAccountRequest {
    string username = 1;
}

message LogoutRequest {
    string username = 1;
}

message LoginRequest {
    string username = 1;
}

message LoginResponse {
    string username = 1;
}

message ListAccountsResponse {
    string username = 1;
    repeated string account = 2;
}

message DeleteAccountResponse {
    // delete account message
    string username = 1;
}

message LogoutResponse {
    // logout message
    string username = 1;
}

// NOT NEEDED
message QuitSession {
    string username = 1;
	string session_id = 2;
}

// NOT NEEDED
message StartSession {
    string username = 1;
}

// NOT NEEDED
message JoinSession {
    string username = 1;
    string session_id = 2;
}

// NOT NEEDED
message StartSessionResponse {
    string session_id = 1;
}

// NOT NEEDED
message JoinSessionResponse {
    string message = 1;
}
